internal/api/mfa.go

```diff
@@ -431,19 +423,14 @@ func (a *API) verifyTOTPFactor(w http.ResponseWriter, r *http.Request, params *V
 	db := a.db.WithContext(ctx)
 	currentIP := utilities.GetIPAddress(r)
 
-	if !factor.IsOwnedBy(user) {
-		// TODO: Should be changed to notFoundError. Retained as internalServerError to preserve backward compatibility.
-		return internalServerError(InvalidFactorOwnerErrorMessage)
-	}
-
 	challenge, err := factor.FindChallengeByID(db, params.ChallengeID)
 	if err != nil && models.IsNotFoundError(err) {
 		return notFoundError(ErrorCodeMFAFactorNotFound, "MFA factor with the provided challenge ID not found")
 	} else if err != nil {
 		return internalServerError("Database error finding Challenge").WithInternalError(err)
 	}
 
-	if challenge.VerifiedAt != nil || challenge.IPAddress != currentIP {
+	if !challenge.IsVerificationIPValid(currentIP) {
```

Comment:
name could be improved, open to suggestions

Reply:
Yeah that makes sense

>on a side note, you can consider moving this check into VerifyFactor along with the FindChallengeByID method and just pass the challenge down in verifyTOTPFactor and verifyPhoneFactor

Yes let's do that - would cut back on duplication by a bit

Reply:
On further thought, I think I might not be able to do that without significant refactoring as [the checks for Verification Enabled](https://github.com/supabase/auth/blob/master/internal/api/mfa.go#L704) probably need to happen before finding a Challenge  / validation so as not to leak information

Will track this on the Notion doc and sit on it for a bit